WordPress/openverse-api
# Remove bespoke throttle exemption

## Fixes


Fixes #585  by @sarayourfriend 

Unblocks https://github.com/WordPress/openverse-frontend/issues/1558

## Description


Removes the throttle exemption classes and switches to just using the scoped OAuth tokens. While that is also bespoke, it already exists and this allows us to delete a substantial amount of hot-path code :slightly_smiling_face:

I wrote some documentation to describe the state of the throttling and authentication world. Please read that, then read this section unless you're already familiar with concepts like rate limit models and throttle scopes (and how they're related).

This PR adds a new rate limit model of "exempt" and ties it to a scope with `None` throttling configuration, which is the same as setting it to infinity. This truly exempts the requester from all rate limiting, just like how the frontend is currently exempting itself via IP address.

In the interest of now having to coordinate a complicated deployment, I've left the IP based throttling in for now.

Before deploying this PR we _MUST_ update the WordPress.com/Jetpack application to have the exempt rate limit model in order to maintain the current behavior of that application.

## Testing Instructions

Note that in addition to below I've also added comprehensive unit tests to the throttling module. If you notice anything missing, please let me know so I can add it.

Please also review the documentation I added and let me know if it needs any clarification.
To test this locally, I recommend using the load testing script from this PR: https://github.com/WordPress/openverse/pull/257

Update settings.py to:
1. Enable throttling
2. Add your docker host address to ALLOWED_HOSTS. This will either be `host.docker.internal` on Docker Desktop or `172.17.0.1` otherwise.

Next create an application and token by following the authentication flow on your local API. Open up Django admin to be able to move the application between the different rate limit models.

I would start by first testing anonymous rate limiting by putting a bogus api key in the load testing script. Confirm that you're quickly rate limited.

Next, use your actual key. Fiddle with the `api.sh` variables of requests and concurrency until you surpass the rate limit. This is testing the "standard" model.

Move the application into the "enhanced" model and do the same. You should be able to overwhelm this throttle by setting the api load testing script to make a little over 200 requests, maybe slightly more due to throttle race conditions. Use 300 and you'll definitely see it rate limit.

Now move the application into the "exempt" model and repeat. You can make the number of requests as high as you want, but you'll never get throttled.

Add your local IP to the `ip-whitelist` in your local redis. I don't know what the best way to find this IP is, it might be different per computer. I just put a breakpoint in a request and check to see what the IP of the request is from inside the API.

Run `just exec cache bash` then `redis-cli`. Add your IP to `ip-whitelist` by running:

```
> sadd "ip-whitelist" "<ip>"
```

Note the quotes are significant here!

Without restarting your API (because the cache is not persistent between restarts on local), make the same anonymous api load testing requests and confirm that you are likewise never throttled.

We will remove this final behavior in a subsequent PR once the frontend has been migrated to use API keys to circumvent the rate limiting.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
